firedrake/mesh.py

```diff
@@ -1091,59 +1090,19 @@ def init_cell_orientations(self, expr):
         if hasattr(self.topology, '_cell_orientations'):
             raise RuntimeError("init_cell_orientations already called, did you mean to do so again?")
 
-        v0 = lambda x: ast.Symbol("v0", (x,))
-        v1 = lambda x: ast.Symbol("v1", (x,))
-        n = lambda x: ast.Symbol("n", (x,))
-        x = lambda x: ast.Symbol("x", (x,))
-        coords = lambda x, y: ast.Symbol("coords", (x, y))
-
-        body = []
-        body += [ast.Decl("double", v(3)) for v in [v0, v1, n, x]]
-        body.append(ast.Decl("double", "dot"))
-        body.append(ast.Assign("dot", 0.0))
-        body.append(ast.Decl("int", "i"))
-
-        # if triangle, use v0 = x1 - x0, v1 = x2 - x0
-        # otherwise, for the various quads, use v0 = x2 - x0, v1 = x1 - x0
-        # recall reference element ordering:
-        # triangle: 2        quad: 1 3
-        #           0 1            0 2
-        if self.ufl_cell() == ufl.Cell('triangle', 3):
-            body.append(ast.For(ast.Assign("i", 0), ast.Less("i", 3), ast.Incr("i", 1),
-                                [ast.Assign(v0("i"), ast.Sub(coords(1, "i"), coords(0, "i"))),
-                                 ast.Assign(v1("i"), ast.Sub(coords(2, "i"), coords(0, "i"))),
-                                 ast.Assign(x("i"), 0.0)]))
-        else:
-            body.append(ast.For(ast.Assign("i", 0), ast.Less("i", 3), ast.Incr("i", 1),
-                                [ast.Assign(v0("i"), ast.Sub(coords(2, "i"), coords(0, "i"))),
-                                 ast.Assign(v1("i"), ast.Sub(coords(1, "i"), coords(0, "i"))),
-                                 ast.Assign(x("i"), 0.0)]))
-
-        # n = v0 x v1
-        body.append(ast.Assign(n(0), ast.Sub(ast.Prod(v0(1), v1(2)), ast.Prod(v0(2), v1(1)))))
-        body.append(ast.Assign(n(1), ast.Sub(ast.Prod(v0(2), v1(0)), ast.Prod(v0(0), v1(2)))))
-        body.append(ast.Assign(n(2), ast.Sub(ast.Prod(v0(0), v1(1)), ast.Prod(v0(1), v1(0)))))
-
-        body.append(ast.For(ast.Assign("i", 0), ast.Less("i", 3), ast.Incr("i", 1),
-                            [ast.Incr(x(j), coords("i", j)) for j in range(3)]))
-
-        body.extend([ast.FlatBlock("dot += (%(x)s) * n[%(i)d];\n" % {"x": x_, "i": i})
-                     for i, x_ in enumerate(expr.code)])
-        body.append(ast.Assign("orientation[0][0]", ast.Ternary(ast.Less("dot", 0), 1, 0)))
-
-        kernel = op2.Kernel(ast.FunDecl("void", "cell_orientations",
-                                        [ast.Decl("int**", "orientation"),
-                                         ast.Decl("double**", "coords")],
-                                        ast.Block(body)),
-                            "cell_orientations")
-
-        # Build the cell orientations as a DG0 field (so that we can
-        # pass it in for facet integrals and the like)
+        if isinstance(expr, expression.Expression):
```

Comment:
Should there be a sanity check here in the case `expr` is just a normal UFL expression?